### DISCUSSION TITLE
社区版 Remix VSCode 开发计划

### DISCUSSION THREAD

**Discussion by drafish** [PlanckerLabs/remix-cc#65] [Ideas]
#### Remix VSCode

remix 官方开发过一个 vscode 插件，但是后来被废弃了，现在 vscode 插件市场已经搜索不到了。我之前有体验过，和 web 版 remix 差距太大了，很难用。

感兴趣的朋友可以把官方仓库 clone 下来，在本地运行这个插件

https://github.com/ethereum/remix-vscode

不过，坑也挺多的，要跑起来并不容易。

#### Remix Desktop

remix 有 desktop ，可以在这里下到

https://github.com/ethereum/remix-desktop/releases

这个 desktop 本质上就是 web 版的 remix ，只不过是用 electron 套了个壳。了解 electron 的朋友都知道，electron 本质上也是个浏览器。

我看过 desktop 的源码，总共也没几行。核心逻辑就是把 web 版 remix 的静态文件压缩包下载下来，然后解压缩，然后再加载。

有一个很坑的点就是，那个静态文件压缩包是从 github 上下载的，见下方链接

https://github.com/ethereum/remix-project/releases

国内用户首次打开 desktop ，大概率会被卡在 github logo 页面。这个时候就是在下载这个压缩包。运气好的话可能卡个把钟头就下载完成了。但你得先退出，然后再打开，才能把页面加载出来。

目前我从 remix 那边了解到，官方打算放弃这个 desktop 的后续维护，转而推出新版的 desktop

#### Remix New Desktop

新版 deskop 目前还在开发中，发布时间待定。官方并没给出详细的信息，只是建了个 milestone 说要做这个事儿。见下方链接

https://github.com/ethereum/remix-project/milestone/6

从 milestone 中的几个任务，可以大致推测出几个比较大的变动

**1. 真的 terminal**

https://github.com/ethereum/remix-project/issues/3721

旧版 desktop 的 terminal 就是 web 版的 terminal ，是个假 terminal 。在浏览器中无法提供真 terminal ，这个大家都理解。但 desktop 也没有，这个有点说不过去了吧。所以这次，remix 新版 desktop 会提供真 terminal 。

**2. 直接访问本地文件**

https://github.com/ethereum/remix-project/issues/3723

这个大家可能有点疑惑，因为旧版 desktop 本来就可以访问本地文件。但我有看过源码，它访问本地文件的方式和 web 版一样，都是要依赖 remixd 。只不过 desktop 帮你自动运行了 remixd 。但是，desktop 不应该本来就有能力访问本地文件的吗。为什么要借助 remixd？我看这个 issue 的意思，应该是要为 desktop 开发一个新版的文件管理系统，可能会实现直接访问本地文件。

**3. 静态文件打包进安装包中**

https://github.com/ethereum/remix-project/issues/3724

这意味着安装完 desktop 后可以直接打开 Remix IDE 。而不用像旧版那样，明明已经下载了安装包，结果装完了是个空壳子，还得下载静态文件压缩包后才能打开。

**补充说明**

以上都是我个人对新版 desktop 的一些猜测，不一定准确。另外，我并不是很看好这个新版的 desktop 。我大概看了下源码，见下方链接

https://github.com/ethereum/remix-project/pull/3885/files

新版 desktop 是和 web 版混在一起了。代码里可以看到很多 `isElectron()` 的判断，就是为了兼容 desktop 和 web 。随着后续的迭代，这种兼容性代码会越来越多，代码会变得越来越臃肿且难以维护。其实现在的 remix 已经很臃肿了，这样一来更是雪上加霜。

## 我为什么看好 VSCode

第一，我自己就用 vscode 。我是做 web 前端开发的，以前也做过两年 nodejs 开发。对于 js 开发者来说，不管是前端还是后端，vscode 都是首选。

第二，智能合约首选语言是 solidity 。而 solidity 生态中的开发工具又大多是基于 js 开发的。所以 js 开发者上手 solidity 会比其他语言的开发者更容易。这就造成了一个结果 —— 虽然 vscode 不是专门为 solidity 打造的，但是有相当一部分人已经在用 vscode 做 solidity 开发了。

第三，remix 有先天缺陷。因为它的文件系统是基于浏览器存储的，你清除了浏览器存储空间，或者浏览器严重崩溃，你的文件就会丢失。虽然你可以用 remixd 来访问本地文件，或者使用 git 把代码保存到远程仓库。但当你这样做的时候，remix 还有那么香吗？vscode 不香吗？

第四，remix 与 vscode 相比最大的优势就是可视化操作界面。但这一点对 vscode 来说，并非无法弥补。就拿编译来说，vscode 的合约编译入口非常隐蔽，没有经验的新人刚开始上手一准会懵逼。但是 vscode 是支持 webview 的，我们可以在 webview 里面加上类似 remix compiler 模块的可视化界面。这样它的易用性就会大大提高了。

第五，抛开开发语言不谈，单从 IDE 的角度来看，vscode 比 remix 更加完善。比如 file explorer、editor、terminal、git 等等，这些功能都比 remix 要做的好。而且 vscode 还集成了云端开发环境，比如他们自家的 github codespaces 。也有别的厂商主动来适配的，比如 nocalhost 。这在 remix 上我感觉就遥遥无期了。

所以，我觉得，既然 vscode 是一个比 remix 更加完善的 IDE ，并且已经有了一定的 solidity 用户基础，为什么我们不把它做的更加好用一点呢。

和 remix 不同的是，IDE 的基础功能我们完全没必要操心，我们只需要关注哪些功能是对 solidity 开发有用的，然后去完善这些功能就可以了。

而且我们都不需要怎么操心功能设计的事儿，因为 remix 已经在 UX 研究上投入了很大的成本，remix 上现有的这些功能，就是他们团队这么多年摸索下来的成果。我们直接照搬过来就可以了。

所以理论上来说，我们完全可以用比 remix 低的多的成本，打造出一款比 remix 用户体验好得多的产品。或许在 web 端不一定能打败 remix ，但是在 desktop 端还是比较稳的，肯定能超越 remix desktop 。

## 为什么 Remix 不选择 VSCode

肯定会有人问这个问题。既然你把 VSCode 说的那么好，为什么 Remix 一开始不选择 VSCode ，甚至现在还放弃了 VSCode 。

我们社区里就有人提过这个问题。为此我还专门去问了 remix 团队的 rob 。他说他们团队16年就开始做 remix 了，那会儿 vscode 还没现在这么火。当时他们就是有个想法，想做个快速验证，基于 web 开发是最快的，所以也没多想就这么干了。后来就一直在这个基础上迭代。等到 vscode 火了以后，remix 也很重了。再往 vscode 上迁移的话，需要重构很多代码。虽然他们后来也做了 vscode 插件，但显然出于成本考虑，并没有花太多时间。但也因为没花太多时间，效果不是很好。到后来干脆就放弃了。

所以在我看来，Remix 放弃 VSCode 插件更多的还是因为历史包袱太重了。但我们没有这个包袱。只要有人有钱，我们完全可以干呀。

#### 一期

先实现三个功能模块，Solidity Compiler 、Deploy & Run 、Debug 。功能设计直接参考 Remix 现有功能。

插件安装参考 Extension Pack for Java 。用户操作的时候只需要安装一个插件，就会自动安装这个插件内包含的多个子插件。

可参考两个仓库

1. remix-vscode：https://github.com/ethereum/remix-vscode
2. remix-light: https://github.com/arthurshoppy/remix-light

我打算招募一个主力开发。最好是对 vscode 和 remix 都比较熟悉且比较有经验的开发者。如果没有同时熟悉 vscode 和 remix 的开发者的话，可以退而求其次，降低对 remix 的要求。因为我对 remix 非常熟悉，我可以辅助开发。

我在这个项目中充当的角色是，项目经理+产品经理+辅助开发。我现在只能制定一个比较粗糙的开发计划。更详细的计划，还得主力开发来和我一起讨论决定。

#### 二期

vscode 本身就集成了云端开发环境，但在 solidity 开发中是否有些痛点没有解决。二期可以考虑解决下这些可能存在的问题。

另外，还可以考虑把 remix 上的插件往 vscode 上搬。有些插件已经有了的，比如 cookbook ，我们可以看下使用起来有什么痛点，然后联系 cookbook 的开发团队怎么合作优化下。

#### 三期

是否会碰到某些功能，需要动 vscode 的源码才能更好地实现。是否会有用户希望不在寄生于 vscode ，而是一个独立的 IDE 。如果能基于 vscode 源码做二次开发的话，我们可以实现一些更加深度的定制化开发。但我不确定是否真的有这样的需求。

**Comment by noyyyy** [comment]
局外人评论一下：

remix 只适合初学者或者写简单的合约。hardhat 有丰富的插件，继承 js 生态，foundry 可以深入 evm 内部做改动。一个刚开始使用 remix 开发的人慢慢到后面会弃用 remix。如果没有熟练的开发人员愿意使用，感觉会很难有社区提 pr 来共同构建。
